### DISCUSSION TITLE
TypeSystem.Format ignored in schema

### DISCUSSION THREAD

**Discussion by enzolupia** [sinclairzx81/typebox#481] [General]
**Typebox version: 0.28.14**

Hi, I'm trying to define a custom format to validate a string header against a custom object schema.
This is a simplified version of what I did so far:

```
const schema = Type.Object(
  {
    client: Type.Object({
      name: Type.String(),
      alias: Type.String()
    })
});

TypeSystem.Format('clientFormat', value => Value.Check(schema, value));

export const clientSchema = Type.String({
  format: 'clientFormat',
});

```

---------------------------------
Finally my request' schema looks like this:

```
export const request = {
  body: Type.Object({
    query: Type.String(),
  }),
  headers: Type.Object({
    'x-client': clientSchema,
  }),
};
```

On server startup I get the following error:
`[serve] FastifyError [Error]: Failed building the validation schema for POST: /api/v2/projects/get-project, due to error unknown format "clientFormat" ignored in schema at path "#/properties/x-client"`

It looks like the custom format is not registered, I didn't  find any additional information in the documentations, am I missing something?

**Comment by sinclairzx81** [comment]
@enzolupia Hi,

The format setup doesn't look quite right. Just a few things 

### Format Registration

The format callback is passed a string, so you need to be checking a string value. The `Value.Check` is testing against an object type.

```typescript
TypeSystem.Format('clientFormat', value => Value.Check(schema, value));
//                                                      ^        ^ value is string
//                                                      |
//                                                      { client: { name: string, alias: string } }

// For example, it should be

TypeSystem.Format('foo-format', value => value === 'foo');
```
### Compiler Setup

Just note if you're using Fastify with the Type Provider, and want to use the TypeBox validation backend, you need to register that with the Fastify instance.

```typescript
import { Type, TypeBoxTypeProvider, TypeBoxValidatorCompiler } from '@fastify/type-provider-typebox'
import Fastify from 'fastify'

const fastify = Fastify().setValidatorCompiler(TypeBoxValidatorCompiler) // set validator compiler here
```

Information on the setup can be found [here](https://github.com/fastify/fastify-type-provider-typebox#type-compiler)

Let me know if that helps
Cheers
S

**Reply by enzolupia** [reply]
Hi @sinclairzx81  of course I forgot to parse the string (it is actually a JSON string) before checking it.

Beside that, I'm already registering TypeBoxValidatorCompiler as validator compiler in Fastify setup as you suggested, yet the custom formats is not being registered.

Thanks for quick response by the way!

**Reply by sinclairzx81** [reply]
@enzolupia Heya,

> Beside that, I'm already registering TypeBoxValidatorCompiler as validator compiler in Fastify setup as you suggested, yet the custom formats is not being registered.

Hmmm, the reason I mention the validatorCompiler setup is that the error message ....

```
[serve] FastifyError [Error]: Failed building the validation schema for POST: /api/v2/projects/get-project, due to error 
unknown format "clientFormat" ignored in schema at path "#/properties/x-client"
```
...looks more like than Ajv error than a TypeBox one (as TB will just return `false` for unknown formats). I'm thinking this might be to do with how Fastify uses Ajv for the `fast-json-stringify` serializer, and that Ajv will probably complain if it doesn't know about the format. Information about adding custom formats to Ajv can be found [here](https://ajv.js.org/guide/formats.html#user-defined-formats), but it would be a bit awkward to have to define the format both for TB and for Ajv. It might be worth testing it to be sure.

The only other advice I can give with respect to TB formats, is that the call to register should be invoked before any validation takes place, but chances are this isn't the issue. 

Again, hope this helps!

**Reply by sinclairzx81** [reply]
@enzolupia Oh, one more thing. The `TypeBoxValidatorCompiler` is actually optional (you don't need to configure it to use the provider). Depending on the challenges getting things to work for both validators, it might be easier to just use the default Ajv compiler instead (as this means you only need to configure the format for one validator instead of both)
